### DISCUSSION TITLE
Error on realtime data fetching: ClientResponseError 0

### DISCUSSION THREAD

**Discussion by lostintheway** [pocketbase/pocketbase#2897] [Q&A] [answered]
I have implemented following code in reactjs for real-time data fetching:

```
  useEffect(() => {
   pb.collection("messages").subscribe<ChatBody>("*", function (data) {
      console.log({ data });
      if (data.action === "create") {
        setMessages((prev) => {
          return [...(prev ?? []), data];
        });
      }
    });
    return () => {
      pb.collection("messages").unsubscribe("*");
    };
  }, []);
```

but it keeps giving me error on page load:
`ClientResponseError.ts:13 Uncaught (in promise) ClientResponseError 0: Something went wrong while processing your request.`

tried adding `pb.autoCancellation(false)` on top but didnt work...  There isn't much more help in documentation, how can i solve this error?

**Comment by ganigeorgiev** [comment]
Is the above executed in a browser or server environment? If the latter, you may need an `EventSource` polyfill, eg.:

```js
// npm install eventsource --save
import eventsource from 'eventsource';

global.EventSource = eventsource;
```
_(for more details, you can check https://github.com/pocketbase/js-sdk#nodejs-via-npm)_

If that's not the case, you can try inspecting the `subscribe` error by chaining a `.catch((err) => { console.log(err.originalError)})` or  `await` the statement and wrap it in `try/catch`. The original error could be accessed via `err.originalError` (see [Error handling](https://github.com/pocketbase/js-sdk#error-handling)).

**Reply by lostintheway** [reply]
I get this error, its executed in a browser. The messaging system is working though.
```
Error: Realtime disconnected.
    at RealtimeService.disconnect (RealtimeService.ts:403:49)
    at RealtimeService.<anonymous> (RealtimeService.ts:110:18)
    at step (tslib.es6.js:147:23)
    at Object.next (tslib.es6.js:150:6)
    at tslib.es6.js:121:71
    at new Promise (<anonymous>)
    at __awaiter (tslib.es6.js:117:12)
    at RealtimeService.unsubscribe (RealtimeService.ts:92:23)
    at RecordService.<anonymous> (RecordService.ts:162:41)
    at step (tslib.es6.js:147:23)
```

**Reply by ganigeorgiev** [reply, marked as answer]
This error is usually thrown when unsubscribing/disconnecting while there is pending connect Promise. 

I guess it could get triggered if the cleanup function (aka. the `unsubscribe` call) in the `useEffect` is triggered before actually being able to establish the realtime connection (aka. the `subscribe` call).

If this behavior is intentional then it is safe to "silent" the error with and empty `.catch((_) => {})` block to your subscribe call.

**Reply by lostintheway** [reply]
thanks... its weird though the cleanup function should only run on component unmounting though, why is it running early? I dont get it, as soon as I load/reload page I get this error....

**Reply by ganigeorgiev** [reply]
_With the planned JS SDK refactoring after v0.17.0 I'll consider resolving the pending promises automatically on disconnect to avoid the above being needed._

**Reply by ganigeorgiev** [reply]
>  its weird though the cleanup function should only run on component unmounting though, why is it running early? 

Without a minimal reproducible repo I'm not sure what could be the reason for the above. Keep in mind that `subscribe` is async call that sends network request(s) and doesn't happen "immediately". If your component is unmounted before the subscribe call being able to complete then I guess the cleanup function will get triggered.

**Reply by ganigeorgiev** [reply]
_@lostintheway Just to let you know that I've submitted a small "fix" in [JS SDK v0.15.3](https://github.com/pocketbase/js-sdk/releases/tag/v0.15.3) that will resolve the pending `subscribe` calls in case `unsubscribe` is called before being able to complete. Note that while this may resolve the above error, make sure that the `useEffect` and its cleanup function are triggered when you expect to avoid burying some other underlying problem (eg. the component is being mounted/unmounted repeatedly/more than once)._
